src/python/pants/python/python_setup.py

```diff
@@ -50,6 +50,24 @@ def register_options(cls, register):
                 "on the format of constraint files and how constraints are applied in Pex and Pip."
             ),
         )
+        register(
+            "--use-all-requirements",
+            advanced=True,
+            # TODO: Should this default to True? Seems like a better initial experience for most
+            #  users (although they still require at least a trivial lockfile to get any benefit).
+            default=False,
+            type=bool,
```

Comment:
To make it easier to introduce multiple lockfiles, I'd recommend factoring these options as "default lockfile"... and that if you set a default lockfile for the repository, it will be used for all targets by default.

That has the advantage of allowing individual targets to override the lockfile, and trigger a separate resolve.

Reply:
I don't think we should optimize too much for this yet. It's very unclear what multiple lockfiles looks like. John raised lots of good, big, unresolved questions like "Can any Python target specify a lockfile?" and "How do we merge the transitive closure?"

Those questions are far too unanswered imo to design this option for the future of multiple lockfiles. Instead, keep it simple for now, and deprecate when we add the new feature.

Reply:
Isn't that orthogonal to this change though? This is just a boolean flag that says "if you can determine a lockfile, use it". We can leave the "how to determine a lockfile to use" part as-is for now? Or am I misunderstanding?

Reply:
> unresolved questions like "Can any Python target specify a lockfile?"

No, only roots. There is prior art for this from other languages that use lockfiles, including rust: https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

> "How do we merge the transitive closure?"

Not precisely sure what you mean here... but I don't think that you would merge the transitive closure. Each distinct lockfile would end up with its own resolve, and there would not be merging across them.

> Isn't that orthogonal to this change though? This is just a boolean flag that says "if you can determine a lockfile, use it". We can leave the "how to determine a lockfile to use" part as-is for now? Or am I misunderstanding?

It's not. The effect would be that "if a lockfile is set in its current location, it is used". IE, setting the existing `[python-setup] requirement_constraints` setting would trigger the behavior, by virtue of declaring a global default lockfile for the repository.

So the reason to think about this now is to avoid having to untangle it (soon, I imagine) in order to support targets needing different resolves.

Reply:
> Not precisely sure what you mean here... but I don't think that you would merge the transitive closure. Each distinct lockfile would end up with its own resolve, and there would not be merging across them.

Target A uses lockfile `one.txt`. Target B depends on A, but also specifies it uses lockfile `two.txt`. What do we do? Indeed, if we only use roots, then this wouldn't be an issue afaict.

Reply:
@stuhood not sure what you meant by "factoring these options as "default lockfile"."

Reply:
"the existing [python-setup] requirement_constraints setting would trigger the behavior, by virtue of declaring a global default lockfile for the repository." seems wrong to me. You can want a lockfile for the benefits it provides (locking down all your versions) while still wanting to do the slow thing and resolve just the deps your target closure needs each time.

Reply:
But I think I'm not understanding what you're suggesting.